### PR TITLE
Spec for Improved keyboard handling in Conpty

### PR DESCRIPTION
**Pull request by zadjii-msft**
## Summary of the Pull Request

We have a number of bugs in the Terminal that all have the same singular root cause - VT input does not carry the same fidelity that Win32 input does. For Win32 applications there are certain keystrokes that simply cannot be represented with VT sequences. 

This is my proposal for how we'll handle all these cases. I'm proposing a _new VT sequence_, which will enable the Terminal to send input with all of the information that an `INPUT_RECORD` might have, to conpty. There, conpty will be able to send input to the client application with the same fidelity they're used to, enabling these keys to work for those applications once again. 

## PR Checklist
* [x] Specs #4999
* [x] I work here
* [x] is a spec

## Detailed Description of the Pull Request / Additional comments
_read the spec_

### CODE REVIEW

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+  union {
+    WCHAR UnicodeChar;
+    CHAR  AsciiChar;
+  } uChar;
+  DWORD dwControlKeyState;
+} KEY_EVENT_RECORD;
+```
+
+To encode all of this information, I propose the following sequence. This is a
+CSI sequence with a final terminator character of `!`. This character appears
```
Comment (j4james) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:138]: Note that `!` is not actually a valid final character for a CSI sequence - it's an intermediate. A standard VT parser would probably expect that to be followed by at least another character. For example, see [`DECSTR`](https://vt100.net/docs/vt510-rm/DECSTR.html).

Comment (zadjii-msft) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md]: Oh, that's a good catch. Thanks! I totally forgot about those. I've update to `_`, which also looks unused.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+
+An application can request `win32-input-mode` with the following OSC sequence
+
+```
+  ^[ ] 1000 ; 1 ; Ps ST
+
+         Ps: Whether to enable or disable win32-input-mode. If omitted, defaults to '0'.
+             0: Disable win32-input-mode
+             1: Enable win32-input-mode
+```
```
Comment (j4james) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:105]: Is there any reason why this isn't just a private `SM`/`RM` mode, the same as `DECKPM`, and Kitty's keymode mode? One advantage is that it can then be queried with `DECRQM` if you need to know whether the mode had actually been applied or not.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+
+       Vk: the value of wVirtualKeyCode - any number. If omitted, defaults to '0'.
+
+       Sc: the value of wVirtualScanCode - any number. If omitted, defaults to '0'.
+
+       Uc: the decimal value of UnicodeChar - for example, NUL is "0", LF is
+           "10", the character 'A' is "65". If omitted, defaults to '0'.
+
+       Cs: the value of dwControlKeyState - any number. If omitted, defaults to '0'.
+```
```
Comment (j4james) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:157]: You may want to consider reordering some of these parameters based on how likely they are to be used. For example, if 99% of the time the repeat count is not needed, then it would make sense to put it last, and then leave it off when not required. Similarly, the control key state is probably going to be 0 a lot of the time too, so that should perhaps be second last. Even keydown will be 0 at least half the time, so that could also potentially be omitted some of the time.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+[VT100.net doc](https://vt100.net/docs/vt510-rm/DECPCTERM.html)
+
+#### Pros:
+* Enables us to send key-down and key-up keys independently
+* Enables us to send modifiers on their own
+* Part of the VT 510 standard
+
+#### Cons:
+* neither `xterm` nor `gnome-terminal` (VTE) seem to implement this. I'm not
+  sure if anyone's got a reference implementation for us to work with.
```
Comment (j4james) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:392]: Not that it matters at this point, but you might be interested to know that Kermit claims to support a PCTERM mode, although it sounds like it needs to be turned on manually (the relevant docs are [here](http://www.columbia.edu/kermit/k95manual/#c7pcterm)). I haven't had a chance to try it out myself, so I don't know how well it works.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+## Potential Issues
+
+_(no change expected)_
+
+## Future considerations
+
+* We could also hypothetically use this same mechanism to send Win32-like mouse
+  events to conpty, since similar to VT keyboard events, VT mouse events don't
+  have the same fidelity that Win32 mouse events do.
+  - We could enable this with a different terminating character, like `"`
```
Comment (j4james) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:305]: I know this is only a future consideration, but note that `"` is not a valid final character. Maybe best to just leave out those details for now.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md (hunk 2 above)
Comment (zadjii-msft) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:105]: I briefly considered it, but I guess I decided against it because I thought it would be easier to not conflict with an existing sequence by using an OSC. I definitely agree thoguh, a private mode does seem more _stylistically_ correct, plus I forgot about the ability to query those modes. 

I'll move this to `CSI ? 9001 h/l`, that doesn't look taken 😄

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md (hunk 3 above)
Comment (zadjii-msft) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:157]: Oh jeez, I forgot about that. That's a really good point. Is it terribly inconsistent to assume that if the repeat count was omitted, it should be `1` instead of `0`? Typically, it looks like key events have `1` there. I'd love to be able to omit it if I can assume missing is 1.

Then, I'd probably make the order `^[ [ Vk ; Sc ; Uc ; Kd ; Cs ; Rc _`. So something like just an 'a' could be 

```
^[[65;30;97;1_
^[[65;30;97_
```

If I can't assume a missing repeat count is 1, then I might as well just stick it at the beginning.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md (hunk 5 above)
Comment (zadjii-msft) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:305]: Good catch, forgot that I put that line in there 😝 thanks!

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md (hunk 3 above)
Comment (j4james) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:163]: I think it's perfectly reasonable to make the default value 1 for the repeat count. Although that's assuming it can never be zero, because it's usually the case that a 0 parameter value would also map to the default value.

Comment (zadjii-msft) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:163]: So, I'm not sure what exactly happens when a repeat count of 0 is passed in. I think internally conhost says "sure buddy, 0, I'm sure what you actually meant was _1_". In my prototype, some keys like Ctrl were still passing a repeat count of 0, but it seemed to work just fine. 

I think it's _probably_ safe though to assume that 0 isn't a valid value and default to 1. Playing with my own console keyboard logger, everything I entered just had a repeat count of 1. If you really want 0, then _don't pass the event at all_...

**Comment by lhecker** [comment]
Uhm I really _really_ don't know if this is any good input but...
I personally feel like for purposes like this a protocol akin to Protobuf (or similar protocols with tagged fields) might be best suited here.

This proposal involves the transmission of 6 positional arguments as the serialized form of a struct.
But technically we're not dealing with C here, but a text protocol instead. As such we're not constrained to something like memory layouts etc., leading to situations like "CreateFile2" being necessary. Positional arguments could lead to exactly such a situation though.

Using tagged fields the wire format might look like this then:
```
  BOOL  bKeyDown;          // => 0
  WORD  wRepeatCount;      // => 1
  WORD  wVirtualKeyCode;   // => 2
  WORD  wVirtualScanCode;  // => 3
  union {
    WCHAR UnicodeChar;     // => 4
    CHAR  AsciiChar;       // => 5
  } uChar;
  DWORD dwControlKeyState; // => 6
} KEY_EVENT_RECORD;

^[[17;29;0;1;8;1_ --> ^[[0:17;1:29;3:1;4:8;5:1_
^[[17;29;0;0;0;1_ --> ^[[0:17;1:29;5:1_
```

Fields that are "empty" (e.g. 0) can be left out.
Fields that are unknown to the receiver can be ignored.
The benefit is the easier addition of new fields and deprecation of old ones.

For instance I'm not super stoked about `UnicodeChar` being a `wchar_t`. It'd be great if we could get rid of surrogate pairs by only transmitting full graphemes. If we decided to implement that we could safely add field "7" and allow supporting applications to support them.

Maybe this idea is totally overkill though, but Protobuf and friends have made my work as backend engineer a lot easier in the past and I kinda felt like this might apply to text protocols too.
It's fine to ignore this. 😄

**Comment by j4james** [comment]
> I personally feel like for purposes like this a protocol akin to Protobuf (or similar protocols with tagged fields) might be best suited here.

I'm not sure what I think about the proposal in general, but I just wanted to point out that our parser doesn't yet support sub-parameters, and that's likely to be a bit of pain to add. We'll probably get there eventually, but I wouldn't want to hold up the keyboard work while we're waiting for that support.

**Comment by lhecker** [comment]
@j4james Aww that's unfortunate. 😞
The parser is implemented in `OutputStateMachineEngine` right? I can see it uses `basic_string_view<size_t>`...
I imagine in the distant future this class might take a text parser instead of the readily parsed parameters. That way it could "pull" the data it needs (even with colons) instead of getting it pushed. Should be easier to extent for such purposes too. 🤔 But... Eh. ¯\\\_(ツ)_/¯

**Comment by jsoref** [comment]
The conflicts shouldn't be hard to resolve `whitelist` was renamed to `expect` -- If you need help, I can help on Sunday.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 6]
```diff
@@ -0,0 +1,500 @@
...
+  including one or more modifiers, an escape sequence encoding the key event is
+  sent
+   - I think like this. ASCII and other keyboard layout chars (things that would
+     hit `SendChar`) would still just come through as the normal char.
+
+#### Cons:
+* Their encoding table is _mad_. [Look at
+  this](https://sw.kovidgoyal.net/kitty/key-encoding.html). What order is that
+  in? Obviously the first column is sorted alphabetically, but the mapping of
+  key->char is in a seemingly bonkers order.
```
Comment (miniksa) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:409]: Might be a bit rude to call it "mad" and "bonkers".

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 7]
```diff
@@ -0,0 +1,500 @@
...
+* I can't get it working locally, so hard to test 😐
+* They do declare the `fullkbd` terminfo capability to identify that they
+  support this mode, but I'm not sure anyone else uses it.
+  - I'm also not sure that any _client_ apps are reading this currently.
+* This isn't designed to be full `KEY_EVENT`s - where would we put the scancode
+  (for apps that think that's important)?
+  - We'd have  to extend this protocol _anyways_
```
Comment (miniksa) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:416]:
```suggestion
  - We'd have to extend this protocol _anyways_

```

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 8]
```diff
@@ -0,0 +1,500 @@
...
+^[[17;29_
+```
+
+* <kbd>Ctrl+Alt+A</kbd>
+```
+^[[17;29;0;1;8_
+^[[18;56;0;1;10_
+^[[65;30;0;1;10_
+^[[65;30;0;0;10_
+^[[18;56;0;0;8_
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:256]: @j4james can correct me if I'm wrong, but the first one might be the typical behavior for something like this. If it is, I'll update the spec to reflect that first even shorter notation.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md (hunk 6 above)
Comment (zadjii-msft) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:409]: Yea that's verbatim from my notes. I'll polish.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 9]
```diff
@@ -0,0 +1,500 @@
...
+```
+
+### Scenarios
+
+#### User is typing into WSL from the Windows Terminal
+
+`WT -> conpty[1] -> wsl`
+
+* Conpty[1] will ask for `win32-input-mode` from the Windows Terminal when
+  conpty[1] first boots up.
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:281]: Oh jeez, yea I missed that. I'll add notes - conpty is going to _always_ ask for this mode from the terminal.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 10]
```diff
@@ -0,0 +1,500 @@
...
+* When the user types keys in Windows Terminal, WT will translate them into
+  win32 sequences and send them to conpty[1]
+* Conpty[1] will translate those win32 sequences into `INPUT_RECORD`s
+* When WSL reads the input, conpty[1] will translate the `INPUT_RECORD`s into VT
```
[the remark this answers is not in the document]
Comment (zadjii-msft) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:285]: Oh jeez, I should clean that up. I thought the `INPUT_RECORD`s were converted into VT-compatible ones when they were _read_ from the buffer, but it looks like they're converted when they're _written_ to the buffer.

**Comment by zadjii-msft** [comment]
regarding:
> I don't know how I feel about the "_" terminator to the CSI. I don't have a better idea, let's say that, but it feels weird to be claiming one. I wish there was something a bit less... scarce... than single-ASCII terminators to be laying claim to. 

I don't love it either, but on the bright side, it seems like as far as input is concerned, there aren't that many in use currently. There's `A`, `B`, `C`, `D`, `H`, `F`, `P`, `Q`, `R`, `S`, (from our own parser) and then I think they decided that wouldn't scale well, and decided to start putting everything else on just `~`. So only those seem used currently.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (miniksa) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md]: I don't really expect the second one to work. It's just a pattern that exists in the world with ipv6 notation. I just wanted you to consider it.

The first one though, I do sort of expect to work.

**Comment by zadjii-msft** [comment]
@j4james You're not _technically_ one of the team members, but as far as our VT support goes, I consider you a core contributor. I want to make sure this definitely gets your sign off as well before I go shipping the implementation PR ☺️

**Comment by j4james** [comment]
I don't like us inventing a proprietary format, but if it has to be done, this seems OK to me. The format is bandwidth heavy, but I suppose that's unavoidable given all the info you want to send. I also have a suspicion that passing an `INPUT_RECORD` from one machine to another might not work as well as you imagine, but that's just a hunch. If it's a problem we'll find out soon enough.

Regarding the `_` final char. it's worth noting that it is actually used in one of the SCO keyboard modes **(<kbd>Ctrl</kbd>+<kbd>Shift</kbd>+<kbd>F10</kbd>** is `CSI _`), so if you're worried about that sort of conflict, then maybe we'll want to come up with something else. Realistically, though, if you want to avoid every possible conflict you'll probably need to add an intermediate char, or otherwise switch the CSI to a DCS or APC. I'm not sure if this is worth worrying about.

**Comment by lhecker** [comment]
In my comment above I mentioned an issue with future compatibility, as I perceived it, for instance if you ever intend to add new fields (or new meaning in general). May I humbly ask if this will be an issue with this new sequence? What would happen if you have an SSH connection and either side sends an older or newer sequence? Does this spec thus (theoretically) require some kind of backward/forward compatibility? Or is that a non-issue? 🙂

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md (hunk 8 above)
[the remark this answers is not in the document]
Comment (j4james) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md]: Missing parameters are interpreted as their default value. So the first example would default the _Uc_ and _Kd_ parameters to 0, set the _Cs_ parameter to 8, and then the _Rc_ parameter that's missing from the end would default to 1.

The second example only skips the one parameter, so _Kd_ becomes 8, and then you have the _Cs_ and _Rc_ missing from the end, which default to 0 and 1 respectively.

**Comment by zadjii-msft** [comment]
>  I also have a suspicion that passing an `INPUT_RECORD` from one machine to another might not work as well as you imagine, but that's just a hunch.

That's a very valid concern. I guess I hadn't considered that the receiving and processing machine might have a different keyboard layout then the sender. 

This sequence is primarily being designed to support the Terminal\<->ConPTY on a local machine scenario. That being said, it definitely could be used by arbitrary client apps as well. At the bare minimum, with the current implementation, there needs to be some sort of disclaimer that these sequences reflect the input from the machine they're being generated on. 

I'll noodle on this for a bit, but I'm not sure how we could go about resolving this. Adding a query/response to ask the terminal for whatever keyboard layout it has? I'm not even sure that's the only thing that would affect the input records generated.

> it is actually used in one of the SCO keyboard modes

Yikes, looking at [this doc](https://vt100.net/docs/vt510-rm/chapter6.html), it looks like they just about claimed _every_ final char in there. I'm not sure this is super worth worrying about. The situation this would be a problem in is if a hypothetical terminal was connected to conpty that sent input to conpty in SCO format. If that terminal was only sending input to conpty in SCO mode, it's going to have much worse problems than just <kbd>Ctrl+Shift+F10</kbd> not working. If we did want to support SCO mode in the future, I'd even go so far as to say we could maybe treat a `win32-input-mode` sequence with no params as <kbd>Ctrl+Shift+F10</kbd>, considering that `KEY_EVENT_RECORD{0}` isn't really valid anyways.

> if you ever intend to add new fields (or new meaning in general).

IMO this is a non-issue - the `INPUT_RECORD` format hasn't changed since before I was born, and there's NO intention by anyone on the console team to make further changes to the Win32 Console API. So I'm not worried about this :)

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 11]
```diff
@@ -0,0 +1,514 @@
...
+> `KEY_EVENT_RECORD`s. It is left as a Future Consideration to add support for
+> the other types of `INPUT_RECORD` as other sequences, which could be done
+> trivially similarly to the following proposal.
+
+## Solution Design
+
+### Inspiration
+
+The design we've settled upon is one that's highly inspired by a few precedents:
+* `Application Cursor Keys (DECCKM)` is a long-supported VT sequences which a
```
Comment (DHowett) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:82]:
```suggestion
* `Application Cursor Keys (DECCKM)` is a long-supported VT sequence which a
```

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 12]
```diff
@@ -0,0 +1,514 @@
...
+  Terminal. This is the DECSET sequence `^[[?1h`/`^[[?1l` (for enable/disable,
+  respectively). This changes the sequences sent by keys like the Arrow keys
+  from a sequence like `^[[A` to `^[OA` instead.
+* The `kitty` terminal emulator uses a similar DECSET sequence for enabling
+  their own input format, which they call ["full mode"]. Similar to DECCKM, this
+  changes the format of the sequences that the terminal should send for keyboard
+  input. Their "full mode" contains much more information when keys are pressed
+  or released (though, less than a full `INPUT_RECORD` worth of data). Instead
+  of input being sent to the client as a CSI or SS3 sequence, this `kitty` mode
+  uses "Application Program-Command" (or "APC) sequences , prefixed with `^[_`.
```
Comment (DHowett) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:93]:
```suggestion
  uses "Application Program-Command" (or "APC") sequences , prefixed with `^[_`.
```

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 13]
```diff
@@ -0,0 +1,514 @@
...
+```
+
+### Scenarios
+
+#### User is typing into WSL from the Windows Terminal
+
+`WT -> conpty[1] -> wsl`
+
+* Conpty[1] will ask for `win32-input-mode` from the Windows Terminal when
+  conpty[1] first boots up. Conpty will _always_ as for win32-input-mode -
```
Comment (DHowett) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:282]:
```suggestion
  conpty[1] first boots up. Conpty will _always_ ask for win32-input-mode -
```

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 14]
```diff
@@ -0,0 +1,514 @@
...
+  Terminals that _don't_ support this mode will ignore this sequence on startup.
+* When the user types keys in Windows Terminal, WT will translate them into
+  win32 sequences and send them to conpty[1]
+* Conpty[1] will translate those win32 sequences into `INPUT_RECORD`s.
+  - When those `INPUT_RECORD`s are written to the input buffer, they'll be
+    converted into VT sequences corresponding to whatever input mode the linux
+    app is in.
```
Comment (DHowett) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:289]: Hey, what happens if a _VT_ application asks for 9001?

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 15]
```diff
@@ -0,0 +1,514 @@
...
+  use to pass a stream of characters to conpty[2].
+* Conpty[2] will get those sequences, and will translate those win32 sequences
+  into `INPUT_RECORD`s
+* When `cmd.exe` reads the input, they'll receive the full `INPUT_RECORD`s
+  they're expecting
+
+
+## UI/UX Design
+
+This is not a user-facing feature.
```
Comment (DHowett) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:324]: Well, not quite. But also _EXTREMELY_.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 16]
```diff
@@ -0,0 +1,514 @@
...
+### Reliability
+
+_(no change expected)_
+
+### Compatibility
+
+This isn't expected to break any existing scenarios. The information that we're
+passing to conpty from the Terminal should strictly have _more_ information in
+them than they used to. Conhost was already capable of translating
+`INPUT_RECORD`s back into VT sequences, so this should work the same as before.
```
Comment (DHowett) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:345]: **must validate**

* [ ] <kbd>Ctrl+H</kbd>
* [ ] <kbd>Ctrl+C</kbd> in the specific instance of #1894
* [ ] **anything requiring <kbd>AltGr</kbd>**
* [ ] Anything as above, but with the <kbd>Ctrl+Alt</kbd> fallback toggle from #6211

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 17]
```diff
@@ -0,0 +1,514 @@
...
+
+There's some hypothetical future where the Terminal isn't connected to conpty.
+In that future, the Terminal will still be able to work correctly, even with
+this ConPTY change. The Terminal will only switch into sending
+`win32-input-mode` sequences when _conpty asks for them_. Otherwise, the
+Terminal will still behave like a normal terminal emulator.
```
Comment (DHowett) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:351]: Add a note here about other terminals that cannot answer 9001 requests.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md (hunk 14 above)
Comment (DHowett) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:289]: I see you addressed this below.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 18]
```diff
@@ -0,0 +1,514 @@
...
+    which type of `INPUT_RECORD` event we're encoding.
+* Client applications that want to be able to read full Win32 keyboard input
+  from `conhost` _using VT_ will also be able to use `^[[?9001h` to do this. If
+  they emit `^[[?9001h`, then conhost will switch itself into
+  `win32-input-mode`, and the client will read `win32-input-mode` encoded
+  sequences as input. This could enable other cross-platform applications to
+  also use win32-like input in the future.
+
+## Options Considered
+
```
Comment (DHowett) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:376]: Options Considered: use an intermediate char - why not?

Comment (DHowett) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:376]: SGR Mouse uses this.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 19]
```diff
@@ -0,0 +1,514 @@
...
+* Can handle solo modifiers
+* Can handle keydown and keyup separately
+* We can make the sequence however we want to parse it.
+
+#### Cons:
+* No reference implementation, so we'd be flying blind
+* We'd be defining our own VT sequences for these, which we've never done
+  before. This was _inevitable_, however, this is still the first time we'd be
+  doing this.
+* Something about Microsoft extending well-defined protocols being bad 😆
```
Comment (DHowett) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:401]: I'd reword this to be a little less... uh, you know.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 20]
```diff
@@ -0,0 +1,514 @@
...
+* By having the Terminal send all input as _this protocol_, VT Input passthrough
+  to apps that want VT input won't work anymore for the Terminal. That's _okay_
```
Comment (DHowett) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:403]: To be fair, we're already breaking VT input passthrough.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 21]
```diff
@@ -0,0 +1,514 @@
...
+
+### kitty extension
+[Reference](https://sw.kovidgoyal.net/kitty/protocol-extensions.html#keyboard-handling)
+
+#### Pros:
+* Not terribly difficult to decode
+* Unique from anything else we'd be processing, as it's an APC sequence
+  (`\x1b_`)
+* From their docs: > All printable key presses without modifier keys are sent
```
Comment (DHowett) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:412]: I don't think `>` works in this context

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 22]
```diff
@@ -0,0 +1,514 @@
...
+  just as in the normal mode. ... For non printable keys and key combinations
+  including one or more modifiers, an escape sequence encoding the key event is
+  sent
+   - I think like this. ASCII and other keyboard layout chars (things that would
```
Comment (DHowett) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:416]: You think like that?

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 23]
```diff
@@ -0,0 +1,514 @@
...
+    CHAR  AsciiChar;
+  } uChar;
+  DWORD dwControlKeyState;
+} KEY_EVENT_RECORD;
+```
+
+To encode all of this information, I propose the following sequence. This is a
+CSI sequence with a final terminator character of `_`. This character appears
+unused as a terminator by sequences _output_ by a client application, and
+doesn't seem to be used as an _input_ sequence terminator either.
```
Comment (j4james) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:146]: Given that we now know `_` is actually used elsewhere, should this bit be updated? Possibly noting why we don't think the overlap is an issue.

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md (hunk 18 above)
Comment (zadjii-msft) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:376]: I suppose we could? I didn't really consider this too strongly one way or another. Since we're not terribly worried about the `_` conflict with the SCO sequences, I figure we might as well not waste another `char` on it, right?

**Comment by zadjii-msft** [comment]
> 
> 
> In case you were waiting for an official approval from me, I have nothing left to complain about. 😄

I wasn't sure the bot would let me request you as a reviewer, but I absolutely _was_ waiting on your official approval on this 😄 I'm glad I have it! Now lets go hope this actually works over in #6309

File: doc/specs/#4999 - Improved keyboard handling in Conpty.md [hunk 24]
```diff
@@ -0,0 +1,559 @@
...
+  union {
+    WCHAR UnicodeChar;
+    CHAR  AsciiChar;
+  } uChar;
+  DWORD dwControlKeyState;
+} KEY_EVENT_RECORD;
+```
+
+To encode all of this information, I propose the following sequence. This is a
+CSI sequence with a final terminator character of `_`. This character appears to
```
Comment (KalleOlaviNiemitalo) [inline doc/specs/#4999 - Improved keyboard handling in Conpty.md:144]: ECMA-48 clause 5.4 lists `_` (05/15) as an unallocated bit combination that is "reserved for future standardization and shall not be used." Did you consider using a Final Byte from the range 07/00 to 07/14, which is available for private use? I see VT510 uses many of those already; perhaps there aren't any left if you want to avoid conflicts with everything and not use Intermediate Bytes.

Comment (j4james) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:144]: That's a good point. Something from the private range would have been more appropriate. But the only character I know that _might_ be unused is `{` (07/11). Since this sequence is only used internally, it's probably not too late to change if anyone feels strongly about it.

Comment (KalleOlaviNiemitalo) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:144]: <https://vt100.net/docs/vt510-rm/chapter4.html> lists two functions with `{` as the Final Byte, but each of those includes an Intermediate Byte.

- Selective Erase Rectangular Area: CSI _Pt_;_Pl_;_Pb_;_Pr_ $ {
- Macro Space Report: CSI _Pn_ * {

Comment (j4james) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:144]: Yeah, there are a quite a few sequences using it with an intermediate. I just meant as a final byte by itself. If we're willing to use an intermediate, there are probably a lot more options available.

Comment (j4james) [reply doc/specs/#4999 - Improved keyboard handling in Conpty.md:144]: I just remembered we discussed this before, and as an input sequence, we found nearly every valid final char was used, so even `{` isn't technically an option if we want to avoid all conflicts. But since this sequence is only used if the win32 input mode is explicitly enabled, we decided it didn't really matter. I suspect nobody will care about it not being private for the same reason.
